Add linked reference tests

coineagan opened the pull request:
These tests focus on note link elements and linked references. They also lay the groundwork for future tests.

Since Notabase can have multiple pages open at the same time, there has to be some way to isolate pages so elements can be reliably targeted. For example, the old `getEditor` `Cypress.Command` used `[data-slate-editor=true]` to target the editor. This works with one page, but if two or more pages were open, or a linked reference existed, then `getEditor` would find more than one `[data-slate-editor=true]` and Cypress would throw an error.

The approach I took is to target individual pages before trying to locate elements within them.

```
cy.targetPage('noteTitle').within(() => {
  // code inside will be isolated to this page
})
```

By isolating each page it removes interference between pages which will make the tests more stable and easy to maintain.

Some things to mention:

I added some data attributes and some of them will be long. For example, I use `data-testid={"note-link-element-" + element.noteTitle}` which would be long if a note’s title is long.

I added a prop to BacklinkNoteBranch so I could create a conditional data attribute: `data-testid={isLinked ? "linked-reference" : "unlinked-reference"}`. But I’m no React expert so it would be nice if you could take a look at that.

There is also an auth issue I have when running all the tests at once, and I’m curious if that is a limitation due to how I hooked up to Supabase or whether you also get the same issue.

churichard commented on cypress/support/commands.js:
```diff
@@ -48,10 +48,62 @@ Cypress.Commands.add(
+// Asserts the amount of references of a specific type a note should have
+// You must pass "linked" or "unlinked" for referenceType
+Cypress.Commands.add(
+  'numberOfReferencesShouldEqual',
```
What do you think about checking the backlink branch title instead (the one that says "X Linked References" or "X Unlinked References"? This will also test that the number displayed is accurate.

This changes the logic such that it'll now be checking the number of references (instead of the number of notes that include a reference), so you'll need to update the tests to account for this. I think this new approach better fits the name of this command, though.

churichard commented on cypress/integration/pages.spec.ts:
```diff
@@ -0,0 +1,49 @@
+
+        // insert returned user_id into '../fixtures/notes.json'
+        for (const note of notes) {
+          ((<any>note).user_id = result.user?.id), data.push(note);
```
Can you remove the use of `any` here?

churichard commented on cypress/support/commands.d.ts:
```diff
@@ -1,20 +1,62 @@
+     * Get linked or unlinked reference based its note title
+     * @example cy.getReference('linked', 'title')
+     */
+    getReference(
```
I actually think it could be easier to read/use if you split this up into `getLinkedReference` and `getUnlinkedReference`

churichard commented on cypress/support/commands.d.ts:
```diff
@@ -1,20 +1,62 @@
+     * Assert there is a certain amount of backlinks
+     * @example cy.numberOfReferencesShouldEqual('linked', 5)
+     */
+    numberOfReferencesShouldEqual(amount: number, referenceType: string);
```
Similar to another comment, I think this could be easier to read/use if you split this up into `numberOfLinkedReferencesShouldEqual` and `numberOfUnlinkedReferencesShouldEqual`

churichard commented on components/sidebar/SidebarNoteLink.tsx:
```diff
@@ -81,7 +81,10 @@ const SidebarNoteLink = (
         </button>
-        <span className="overflow-hidden overflow-ellipsis whitespace-nowrap">
+        <span
+          data-testid={'sidebar-note-link-' + note.title}
```
Instead of adding a `data-testid` here, can we get this element by using `contains`?

churichard commented on components/editor/elements/NoteLinkElement.tsx:
```diff
@@ -32,6 +32,7 @@ export default function NoteLinkElement(props: NoteLinkElementProps) {
     <Tooltip content={element.noteTitle} placement="bottom-start">
       <span
         role="button"
+        data-testid={'note-link-element-' + element.noteTitle}
```
Instead of adding a `data-testid` here, can we get this element by using `contains`? I want to avoid using the note title in the `data-testid` if possible, because it's user-controlled and could have spaces or be very long.

When using `contains`, the displayed note name could be either the note title or some custom text that the user specifies, but I think that should be okay. We'll just need to account for that when using the command.

churichard reviewed (changes requested):
Overall looks very solid! I tested locally and all of these tests pass for me (no auth issues on my end).

I left some comments that would be good to address before landing this.

coineagan commented:
Thanks for taking a look at these! I think your suggestions would all be great changes. All of your comments were things I wanted to ask your opinion about, so I’m happy you picked up on them. Even the ones I forgot to ask!

I will make the changes probably this weekend or early next week.

coineagan commented:
I made the changes you suggested and added an additional test to count the total number of linked references.

To separate the finding of linked references and the number of notes with linked references, I removed `numberOfReferencesShouldEqual` and created `getNumberOfNotesWithLinkedReferences` and `getNumberOfLinkedReferencesTo`.

churichard commented on cypress/support/commands.js:
```diff
@@ -75,27 +77,22 @@ Cypress.Commands.add('getNoteTitle', (noteTitle) => {
+});
+
+// Gets the number of linked references to a page
+Cypress.Commands.add('getNumberOfLinkedReferencesTo', (pageTitle) => {
```
I think you may have forgotten to use the `pageTitle` parameter here.

churichard reviewed (commented):
Thanks for implementing my suggestions!

Could you rebase onto the `main` branch? It looks like there may be a merge conflict.

coineagan commented:
I reworked the `getNumberOfLinkedReferencesTo` function in the [new commit](https://github.com/churichard/notabase/pull/1227/commits/e7d59a8437ed70662dee2b29ceab98ac6e722318)  to not need the pageTitle argument.

churichard reviewed (approved):
This looks awesome! Thanks again for adding these.
